ASM-6587 Workaround NoMethodError in toRbvmomiTypeHash method in rbvmomi

kaushalgala opened the pull request:
Seems with ESX 6.0, VMware added some APIs which do not have the returnTypeInfo defined in VMODL:
e.g. WitnessJoinVsanCluster, WitnessSetPreferredFaultDomain, AddUnicastAgentWitness, LeaveVsanCluster, RemoveUnicastAgent

The EsxCliNamespace used by the host object in ESX discovery and software updates ends up invoking RbVmomi::VIM::DynamicTypeMgrManagedTypeInfo.toRbvmomiTypeHash, where it tries to access properties of returnTypeInfo of every VMODL APIs.

As a result, when it encounters any of above mentioned methods, it results into NoMethodError.

To workaround such problem, we monkey-patch the toRbvmomiTypeHash by setting result = nil for such methods without any returnTypeInfo.

kaushalgala commented:
@gavin-scott , @ripienaar , @sushilrai  please review this, and suggest if there is better alternative to workaround rbvmomi bugs.

kaushalgala commented:
@asaenz, review this, and suggest if there is better alternative to workaround rbvmomi bugs.

sushilrai commented on lib/puppet/provider/esx_software_update/default.rb:
```diff
@@ -1,6 +1,7 @@
 provider_path = Pathname.new(__FILE__).parent.parent
 require File.join(provider_path, 'vcenter')
 require 'rbvmomi'
+require File.join(provider_path, 'custom_rbvmomi')
```
Looks good for addressing missing result in the return-type. Few questions:

- Is there a PR in upstream vmware-vcenter repo that we can pull which has this fix? 
- As per the description, return value is missing in multiple APIs. Instead of including the custom rbvmomi in just this provider, can we include this in transport layer so that all APIs can be addressed.

kaushalgala replied:
Thanks Sushil for feedback.

Answer to question 1:
I checked in upstream vmware-vcenter and didn't find any PR we can pull in.  In fact, the right thing would be to have public rbvmomi itself fixed to account for missing result-type.  We can put in a PR for rbvmomi, and once they accept it and push out new rbvmomi gem, we can get rid of our custom_rbvmomi patch.

Answer to question 2:
Yes, I can put this in parent class 'vcenter.rb'', so that all providers get this patch.. Will send an updated PR.

kaushalgala commented:
@sushilrai , have addressed your feedback - now this rbvmomi override is used in transport, and hence available to all providers.

if this looks good to you, feel free to merge the PR

sushilrai commented on bin/esx_software_discovery.rb:
```diff
@@ -1,6 +1,7 @@
 #!/opt/puppet/bin/ruby
 require 'json'
 require 'rbvmomi'
+require_relative '../lib/puppet_x/puppetlabs/transport/rbvmomi_patch' # Use patched library to workaround rbvmomi issues
```
do we still need this. changes added to vmware.rb will be sufficient

kaushalgala replied:
esx_software_discovery.rb does not use any of puppet modules, including vmware.rb.  So we need it here.
